OpenSimRT/RealTime/include/RealTimeAnalysis.h

```diff
@@ -0,0 +1,234 @@
+/**
+ * -----------------------------------------------------------------------------
+ * Copyright 2019-2021 OpenSimRT developers.
+ *
+ * This file is part of OpenSimRT.
+ *
+ * OpenSimRT is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, either version 3 of the License, or (at your option) any later
+ * version.
+ *
+ * OpenSimRT is distributed in the hope that it will be useful, but WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+ * A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * OpenSimRT. If not, see <https://www.gnu.org/licenses/>.
+ * -----------------------------------------------------------------------------
+ *
+ * @file RealTimeAnalysis.h
+ *
+ * @brief This file contains a facade class as a convinient interface for
+ * performing musculoskeletal analysis in mocap scenarios.
+ *
+ * @author Dimitar Stanev <jimstanev@gmail.com>
+ * contribution: Filip Konstantinos <filip.k@ece.upatras.gr>
+ */
+#pragma once
+
+#include "CircularBuffer.h"
+#include "OpenSimUtils.h"
+#include "SignalProcessing.h"
+#include "InverseDynamics.h"
+#include "InverseKinematics.h"
+#include "MuscleOptimization.h"
+#include "JointReaction.h"
+#include "RealTimeAnalysis.h"
+#include "internal/RealTimeExports.h"
+
+#include <atomic>
+
+namespace OpenSimRT {
+/**
+ * @brief Convinient struct container of the marker/orientation data and the
+ * external measured wrenches in Mocap scenarios.
+ */
+struct RealTime_API MotionCaptureInput {
+    InverseKinematics::Input IkFrame;
+    std::vector<ExternalWrench::Input> ExternalWrenches;
+};
+
+/**
+ * @brief A higher order function that is an interface between the motion
+ * capture system and the InverseKinematics input. Termination of the
+ * acquisition should be handled by throwing an exception.
+ */
+typedef std::function<MotionCaptureInput()> DataAcquisitionFunction;
+
+/**
+ * @brief Provides a convinient interface for performing RT musculoskeletal
+ * analysis. It creates one thread for the data acquisition, IK and filtering,
+ * and one processing thread for the rest the analysis (ID, SO and JR).
+ */
+class RealTime_API RealTimeAnalysis {
+ public:
+    struct UnfilteredData {
```

Comment:
This was implemented initially due to the time constraints of the review. I was thinking that it would be best if we could view this as a continuous array of values or separately to avoid converting the vectors. Maybe a union or a view can work. Do you have any other ideas?

Reply:
E.g., to be able to access t, q, w or get the whole thing as a single array without any additional computational cost.

Reply:
Since the filter requires a `SimTK::Vector` as input, the only thing I can think of is to remove the `UnfilteredData` struct entirely and keep just a function identical to `RealTimeAnalysis::UnfilteredData::toVector()` to create the input vector. This will remove the intermediate step of instantiating `UnfilteredData` objects in every loop, copying data to the struct fields and then applying `RealTimeAnalysis::UnfilteredData::toVector()`.